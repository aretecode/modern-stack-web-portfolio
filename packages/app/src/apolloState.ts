// tslint:disable typedef
// @lint: ^ this is because we have typings, no need to define for each
import { gql } from 'apollo-boost'
import { withClientState } from 'apollo-link-state'
import { cache } from './apolloCache'
import { logger } from './log'
import { Resolvers, ResumeType } from './typings'
import { addTypeName } from './utils/addTypeName'
import ResumeQuery from './graphql/Resume'

export const typeDefs = gql`
  type Profile {
    network: string
    username: string
    url: string
  }
  type Basics {
    name: string
    label: string
    picture: string
    email: string
    telephone: string
    website: string
    summary: string
    address: string
    postalCode: string
    city: string
    countryCode: string
    region: string
    profiles: [Profile]
    resumeWebsite: string
    skills: [string]
  }
  type Work {
    company: string
    position: string
    startDate: string
    endDate: string
    summary: string
    highlights: [string]
    website: string
    picture: string
  }
  type Resume {
    id: ID
    basics: Basics
    work: [Work]
  }

  type AddOrUpdateResumeResponse {
    responseMessage: string
  }

  type Query {
    resume(id: ID): [Resume]
  }
  type Mutation {
    setResume(basics: Basics, work: [Work]): AddOrUpdateResumeResponse!
  }
`

export const apolloState = {
  defaults: {
    // @todo WithTypeNameStrings<ResumeType>
    resume: {
      __typename: 'Resume',
      basics: {
        __typename: 'Basics',
        name: '',
        label: '',
        picture: '',
        email: '',
        telephone: '',
        website: '',
        summary: '',
        profiles: [],
        address: '',
        postalCode: '',
        city: '',
        countryCode: '',
        region: '',
        resumeWebsite: '',
        skills: [],
      },
      work: [],
    } as any,
  },
  resolvers: {
    Query: {
      async resume(obj, args, context, info) {
        logger.info('[query] resume')

        const data = context.cache.read<{ resume: ResumeType }>({
          query: ResumeQuery,
          optimistic: true,
        })

        logger.log({ data })

        return { ...data!.resume }
      },
    },
    Mutation: {
      /**
       * @todo generate typings from graphql schema & import
       * could flatten it out
       */
      async setResume(
        obj,
        args: { id?: string; basics: { profiles?: {} }; work: unknown[] },
        context,
        info
      ) {
        logger.info('[mutation] setResume')
        logger.dir(args)

        const updated = {
          __typename: 'Resume',
          basics: {
            __typename: 'Basics',
            ...args.basics,
            profiles: addTypeName('Profile', args.basics.profiles),
          },
          work: addTypeName('Work', args.work),
        }
        logger.info('[mutation] updated')
        logger.info(updated)

        context.cache.writeData({
          data: {
            resume: { ...updated },
            ...updated,
          },
        })

        return {
          __typename: 'AddOrUpdateResumeResponse',
          responseMessage: 'Success if it does not throw?',
        }
      },
    },
  } as Resolvers,
}

export const stateLink = withClientState({
  cache,
  typeDefs,
  ...apolloState,
})
