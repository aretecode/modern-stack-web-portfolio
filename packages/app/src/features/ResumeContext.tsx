/**
 * can split ResumeContext & ResumeProvider
 */
import * as React from 'react'
import { Query } from 'react-apollo'
import { EMPTY_OBJ, EMPTY_ARRAY } from '../utils/EMPTY'
import ResumeQuery from '../graphql/Resume'
import { ResumeResponse, GraphqlProps, BasicsType, WorkType } from '../typings'

export interface ResumeContextType {
  isLoading?: boolean
  basics: BasicsType
  work: WorkType[]
}

export const EMPTY_RESUME_CONTEXT = Object.freeze({
  basics: EMPTY_OBJ,
  work: EMPTY_ARRAY,
}) as ResumeContextType

export const ResumeContext = React.createContext<ResumeContextType>(
  EMPTY_RESUME_CONTEXT
)

export class ResumeProvider extends React.PureComponent {
  renderContext = (response: GraphqlProps<ResumeResponse>) => {
    /**
     * @todo @@perf can simplify & improve
     */
    const data = { resume: EMPTY_ARRAY, ...response.data! }
    const { refetch, loading } = response
    const resume = {
      basics: data.resume.basics || {
        profiles: EMPTY_ARRAY,
      },
      work: data.resume.work || EMPTY_ARRAY,
    }
    const value = { isLoading: !!loading, refetch, ...resume }

    return (
      <ResumeContext.Provider value={value}>
        {this.props.children}
      </ResumeContext.Provider>
    )
  }

  render() {
    return (
      <Query<ResumeResponse> query={ResumeQuery} pollInterval={5000}>
        {this.renderContext}
      </Query>
    )
  }
}
