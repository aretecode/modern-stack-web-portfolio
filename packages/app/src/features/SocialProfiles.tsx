import * as React from 'react'

export function SocialProfiles() {
  return (
    <img src="https://user-images.githubusercontent.com/4022631/55691849-01c82e80-59c0-11e9-8c5b-69f64a055b1f.png" />
  )
}
