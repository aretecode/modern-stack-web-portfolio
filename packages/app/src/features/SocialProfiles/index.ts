export { default } from './SocialProfiles'
export { default as SocialProfiles } from './SocialProfiles'
export * from './SocialProfileIcon'
