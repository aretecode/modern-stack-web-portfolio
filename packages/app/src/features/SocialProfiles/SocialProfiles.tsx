import * as React from 'react'
import styled from 'styled-components'
import { ProfileType } from '../../typings'
import { ResumeContext, ResumeContextType } from '../ResumeContext'
import { StyledLink } from '../Link'
import { SocialProfileIcon, IconNameType } from './SocialProfileIcon'

/**
 * @todo material-ui timings
 * @todo when transitioning mobile to desktop, needs to morph
 */
export const StyledSocialProfilesWrap = styled.aside`
  display: flex;
  padding: 1rem;
  justify-content: space-between;
  flex-direction: column;
  transition: flex-direction 0.24s ease-in-out, width 0.24s ease-in-out;

  @media (max-width: 1023px) {
    width: 100%;
    flex-direction: row;
    order: 1;
    border-top: 4px solid;
    margin-top: 1rem;
  }
`

export const StyledSocialProfileIcon = styled(SocialProfileIcon)`
  width: 2rem;
  height: 2rem;
`

export function renderProfileItem(profile: ProfileType, index?: number) {
  return (
    <StyledLink
      rel="me"
      href={profile.url}
      title={profile.username}
      key={profile.url}
    >
      <StyledSocialProfileIcon icon={profile.network} />
    </StyledLink>
  )
}

/**
 * @poc
 *    <img src="https://user-images.githubusercontent.com/4022631/55691849-01c82e80-59c0-11e9-8c5b-69f64a055b1f.png" />
 */
export default class SocialProfiles extends React.PureComponent {
  static contextType = ResumeContext
  readonly context: ResumeContextType

  render() {
    return (
      <StyledSocialProfilesWrap>
        {this.context.basics.profiles.map(renderProfileItem)}
        {renderProfileItem({
          url: this.context.basics.resumeWebsite,
          username: 'download resume',
          network: 'pdf',
        })}
      </StyledSocialProfilesWrap>
    )
  }
}
