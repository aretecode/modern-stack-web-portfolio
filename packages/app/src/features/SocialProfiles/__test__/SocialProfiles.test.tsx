import * as React from 'react'
import { render } from 'react-testing-library'
import { ResumeContext } from '../../ResumeContext'
import { SocialProfileIcon } from '../SocialProfileIcon'
import SocialProfiles from '../SocialProfiles'
import { defaultApolloStateResume } from '../../../apolloState'

describe('SocialProfiles', () => {
  it('should render icon for name', () => {
    const { container } = render(<SocialProfileIcon icon="pdf" />)
    expect(container).toMatchSnapshot()
  })
  it('should render all for context', () => {
    const { container } = render(
      <ResumeContext.Provider value={defaultApolloStateResume}>
        <SocialProfiles />
      </ResumeContext.Provider>
    )
    expect(container).toMatchSnapshot()
  })
})
