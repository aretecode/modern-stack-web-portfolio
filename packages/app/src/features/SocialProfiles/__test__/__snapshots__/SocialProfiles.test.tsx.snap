// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialProfiles should render all for context 1`] = `
.c1 {
  -webkit-text-decoration: none;
  text-decoration: none;
  position: relative;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  color: var(--color-link);
  padding-bottom: 2px;
}

.c1:link,
.c1:visited {
  color: var(--color-link);
}

.c1:after {
  display: block;
  position: absolute;
  content: '\\0020';
  z-index: 2;
  border-bottom: 3px solid var(--color-link);
  right: 0;
  bottom: -3px;
  left: 0;
  -webkit-transition: border-bottom-width 0.3s;
  transition: border-bottom-width 0.3s;
}

.c1:hover:after {
  border-bottom-width: 0.5rem;
}

.c1:focus {
  outline: thin dotted;
}

.c1:hover {
  color: #aaa;
}

.c1:link {
  -webkit-tap-highlight-color: rgba(102,102,102,0.5);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 1rem;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transition: flex-direction 0.24s ease-in-out,width 0.24s ease-in-out;
  transition: flex-direction 0.24s ease-in-out,width 0.24s ease-in-out;
}

.c2 {
  width: 2rem;
  height: 2rem;
}

@media (max-width:1023px) {
  .c0 {
    width: 100%;
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    -webkit-order: 1;
    -ms-flex-order: 1;
    order: 1;
    border-top: 4px solid;
    margin-top: 1rem;
  }
}

<div>
  <aside
    class="c0"
  >
    <a
      class="c1"
      href="https://www.linkedin.com/in/james-wiens/"
      rel="me"
      title="aretecode"
    >
      <svg
        class="c2"
        role="img"
        viewBox="0 0 24 24"
        xmlns="https://www.w3.org/2000/svg"
      >
        <title>
          linkedin
           icon
        </title>
        <path
          d="M16.893 20.452v-5.569c0-1.328-.027-3.037-1.852-3.037-1.853 0-2.136 1.445-2.136 2.939v5.667H9.351V9h3.414v1.561h.046c.477-.9 1.637-1.85 3.37-1.85 3.601 0 4.267 2.37 4.267 5.455v6.286h-3.555zM5.337 7.433a2.062 2.062 0 0 1-2.063-2.065 2.064 2.064 0 1 1 2.063 2.065zm1.782 13.019H3.555V9h3.564v11.452z"
          fill="#000"
        />
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
      </svg>
    </a>
  </aside>
</div>
`;

exports[`SocialProfiles should render icon for name 1`] = `
<div>
  <svg
    role="img"
    viewBox="0 0 24 24"
    xmlns="https://www.w3.org/2000/svg"
  >
    <title>
      pdf
       icon
    </title>
    <path
      d="M20 2H8c-1.1 0-2 .9-2 2v12c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-8.5 7.5c0 .83-.67 1.5-1.5 1.5H9v1.25c0 .41-.34.75-.75.75s-.75-.34-.75-.75V8c0-.55.45-1 1-1H10c.83 0 1.5.67 1.5 1.5v1zm5 2c0 .83-.67 1.5-1.5 1.5h-2c-.28 0-.5-.22-.5-.5v-5c0-.28.22-.5.5-.5h2c.83 0 1.5.67 1.5 1.5v3zm4-3.75c0 .41-.34.75-.75.75H19v1h.75c.41 0 .75.34.75.75s-.34.75-.75.75H19v1.25c0 .41-.34.75-.75.75s-.75-.34-.75-.75V8c0-.55.45-1 1-1h1.25c.41 0 .75.34.75.75zM9 9.5h1v-1H9v1zM3 6c-.55 0-1 .45-1 1v13c0 1.1.9 2 2 2h13c.55 0 1-.45 1-1s-.45-1-1-1H5c-.55 0-1-.45-1-1V7c0-.55-.45-1-1-1zm11 5.5h1v-3h-1v3z"
    />
    <path
      d="M0 0h24v24H0V0z"
      fill="none"
    />
  </svg>
</div>
`;
